### PR TITLE
Remove SpectrumFit class

### PR DESCRIPTION
**Pull request by adonath**
This PR removes the `SpectrumFit` and `SpectrumFitResult` class from Gammapy. It contains
 the following changes:

- Remove `SpectrumFit` from the Gammapy code base, tutorials and docs
- Remove `SpectrumFitResult` object. The I/O functionality was moved to `SpectrumAnalysisIACT.write()`, which was the only placed it was used.
- For the transition phase I implemented a `SpectrumObservationList.to_spectrum_datasets()` method, which is used in a few places.
- Remove test duplications in `spectrum/test/test_fit.py` and `spectrum/test/test_dataset.py`

### CODE REVIEW

File: docs/spectrum/fitting.rst [hunk 1]
```diff
@@ -17,75 +17,73 @@ Getting Started
...
+    >>> print(model)
 
-    Parameters:
+        PowerLaw
 
-           name     value     error         unit         min    max
-        --------- --------- --------- --------------- --------- ---
-            index 2.761e+00 1.094e-01                       nan nan
-        amplitude 5.118e-11 4.849e-12 1 / (cm2 s TeV)       nan nan
-        reference 1.000e+00 0.000e+00             TeV 0.000e+00 nan
```
Comment (adonath) [inline docs/spectrum/fitting.rst]: Note, that the numbers here changed for the sherpa example (see below) equivalently. So the different numbers are likely to a change in the input data and not because of the `SpectrumFit` -> `Fit` transition.

File: gammapy/scripts/spectrum_pipe.py [hunk 2]
```diff
@@ -73,30 +72,74 @@ def run_extraction(self):
...
+            outfile.write(val)
+
     def run_fit(self, optimize_opts=None):
         """Run all step for the spectrum fit."""
-        self.fit = SpectrumFit(
-            obs_list=self.extraction.spectrum_observations, **self.config["fit"]
-        )
-        self.fit.run(optimize_opts=optimize_opts)
-        modelname = self.fit.result[0].model.__class__.__name__
+        datasets_fit = self.extraction.spectrum_observations.to_spectrum_datasets(**self.config["fit"])
```
Comment (registerrier) [inline gammapy/scripts/spectrum_pipe.py]: I suppose we can keep like this for the moment, but this will likely have to significantly change if `SpectrumObservation` and `SpectrumObservationList` are removed.

I suppose that a loop on all `Dataset` to set the model is unavoidable step for the user.

File: gammapy/spectrum/observation.py [hunk 3]
```diff
@@ -705,6 +718,39 @@ def show_obs(idx):
         return interact(show_obs, idx=(0, max_, 1))
 
 
+    def to_spectrum_datasets(self, model=None, fit_range=None, forward_folded=True):
```
Comment (registerrier) [inline gammapy/spectrum/observation.py]: This will likely go if `SpectrumExtract` no longer returns a `SpectrumObservationList` but simply a list of `SpectrumDatasetOnOff`.

File: tutorials/spectrum_pipe.ipynb [hunk 4]
```diff
@@ -96,7 +96,7 @@
     "    reference=1 * u.TeV,\n",
     ")\n",
     "\n",
-    "flux_point_binning = EnergyBounds.equal_log_spacing(0.7, 30, 5, u.TeV)"
+    "flux_point_binning = EnergyBounds.equal_log_spacing(0.77, 30, 5, u.TeV)"
```
Comment (registerrier) [inline tutorials/spectrum_pipe.ipynb]: why the change?

File: tutorials/hess.ipynb [hunk 5]
```diff
@@ -601,9 +637,17 @@
    "source": [
     "## Exercises\n",
     "\n",
+    "- Turn of the containment correction of the 1D spectral analysis and compare the result to the reference Crab spectrum\n",
```
Comment (registerrier) [inline tutorials/hess.ipynb]: Turn off

File: tutorials/hess.ipynb [hunk 6]
```diff
@@ -323,7 +324,42 @@
...
+    "\n",
+    "ax_spectrum.set_ylim(1e-11, 1e-10)\n",
+    "ax_spectrum.legend();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The difference we see compared to the reference H.E.S.S. 2006 model is because of the appilcation of the containment correction. "
```
Comment (registerrier) [inline tutorials/hess.ipynb]: application

File: tutorials/spectrum_pipe.ipynb [hunk 7]
```diff
@@ -96,7 +96,7 @@
     "    reference=1 * u.TeV,\n",
```
Comment (registerrier) [inline tutorials/spectrum_pipe.ipynb]: There is `SkyCoord.from_name("crab")` that breaks Travis CI. Maybe this should be removed?

File: tutorials/hess.ipynb (hunk 6 above)
Comment (registerrier) [reply tutorials/hess.ipynb]: The difference should not come from containment correction. It is also applied in the 2006 Crab paper.

File: gammapy/spectrum/tests/test_fit.py [hunk 8]
```diff
@@ -152,47 +123,6 @@ def setup(self):
         )
 
 
-    def test_basic_results(self):
```
Comment (registerrier) [inline gammapy/spectrum/tests/test_fit.py]: All those tests are replaced by tests in test_dataset.py, right?

Comment (adonath) [reply gammapy/spectrum/tests/test_fit.py]: Yes, I removed the duplication of tests between `test_fit.py` and `test_dataset.py`. The distinction is a bit arbitrary, but for now it shouldn't matter. I guess it's OK to clean up the tests in future.

File: gammapy/scripts/spectrum_pipe.py (hunk 2 above)
Comment (adonath) [reply gammapy/scripts/spectrum_pipe.py]: Yes, this is only meant to be a preliminary solution to help with the transition from `SpectrumObservation` to `SpectrumDataset`. I've opened a reminder issue #2113 to discuss the API question how to set models on `Datasets`. Maybe keeping the loop over dataset objects is just fine...

File: gammapy/spectrum/observation.py [hunk 9]
```diff
@@ -518,22 +518,35 @@ def copy(self):
         """A deep copy."""
         return copy.deepcopy(self)
 
-    def to_spectrum_dataset(self):
+    def to_spectrum_dataset(self, model=None):
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/spectrum/observation.py]: As I've written below it's just meant to be a preliminary solution to help with the transition from `SpectrumObservation` to `SpectrumDatasetOnOff`. I just thought it was better to have this transition layer in one single place. But thinking about it again it's probably better to set the model afterwards, because that's what we have to do for the `SpectrumDatasetOnOff` anyway...I guess this is what you meant?

File: docs/spectrum/fitting.rst [hunk 10]
```diff
@@ -17,75 +17,73 @@ Getting Started
...
-    FitResult
+
+        OptimizeResult
 
         backend    : minuit
         method     : minuit
         success    : True
-        nfev       : 99
-        total stat : 63.00
+        nfev       : 115
```
[the remark this answers is not in the document]
Comment (adonath) [reply docs/spectrum/fitting.rst]: I was also surprised by this change, because in the tests not a single number changed. Interestingly the sherpa example in the docs changed as well and also gave different, but consistent numbers with the other example. So I think we probably forgot to update this docs example to the latest version of the `hess-dl3` dataset. The change is definitely not related to the `SpectrumFit` -> `Fit` transition.

File: gammapy/spectrum/flux_point.py [hunk 11]
```diff
@@ -793,7 +793,10 @@ def __init__(
         sigma_ul=2,
     ):
         # make a copy to not modify the input datasets
-        datasets = Datasets(datasets).copy()
+        if not isinstance(datasets, Datasets):
+            datasets = Datasets(datasets)
+
+        datasets = datasets.copy()
 
         if not datasets.is_all_same_type and datasets.is_all_same_shape:
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/spectrum/flux_point.py]: Done

File: tutorials/hess.ipynb (hunk 6 above)
Comment (adonath) [reply tutorials/hess.ipynb]: As a quick check I found that the un-corrected spectrum is much closer to the 2006 crab reference, but as I didn't check in detail it's probably better to remove the statement again.

File: tutorials/spectrum_pipe.ipynb (hunk 7 above)
Comment (adonath) [reply tutorials/spectrum_pipe.ipynb]: Done

File: tutorials/spectrum_pipe.ipynb (hunk 4 above)
Comment (adonath) [reply tutorials/spectrum_pipe.ipynb]: This change was not intended, I'll revert it.

File: gammapy/spectrum/observation.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/spectrum/observation.py]: I've looked into this again and the problem was that `SpectrumDataset` does not allow `model=None` on init. But it's probably the much better idea to change it there...

**Comment by adonath** [comment]
The remaining Travis-CI fail is unrelated. I'll merge this PR now.
